packages/core/contracts/polygon/OracleChildTunnel.sol

```diff
@@ -0,0 +1,117 @@
+     * @param sender The sender of `data` from the Root chain.
+     * @param data ABI encoded params with which to call `_publishPrice`.
+     */
+    function _processMessageFromRoot(
+        uint256, /* stateId */
+        address sender,
+        bytes memory data
+    ) internal override validateSender(sender) {
+        (bytes32 identifier, uint256 time, bytes memory ancillaryData, int256 price) =
+            abi.decode(data, (bytes32, uint256, bytes, int256));
+        _publishPrice(identifier, time, ancillaryData, price);
+    }
+
+    /**
+     * @notice Returns whether a price has resolved for the request.
+     * @param identifier Identifier of price request.
+     * @param time Timestamp of price request
+     * @param ancillaryData extra data of price request.
+     * @return True if a price is available, False otherwise. If true, then getPrice will succeed for the request.
+     */
+    function hasPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public view override onlyRegisteredContract() returns (bool) {
+        bytes32 priceRequestId = _encodePriceRequest(identifier, time, _stampAncillaryData(ancillaryData, msg.sender));
+        return prices[priceRequestId].state == RequestState.Resolved;
+    }
+
+    /**
+     * @notice Returns resolved price for the request.
+     * @dev Reverts if price is not available.
+     * @param identifier Identifier of price request.
+     * @param time Timestamp of price request
+     * @param ancillaryData extra data of price request.
+     * @return int256 Price, or reverts if no resolved price for any reason.
+     */
+    function getPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public view override onlyRegisteredContract() returns (int256) {
+        bytes32 priceRequestId = _encodePriceRequest(identifier, time, _stampAncillaryData(ancillaryData, msg.sender));
+        Price storage lookup = prices[priceRequestId];
+        require(lookup.state == RequestState.Resolved, "Price has not been resolved");
+        return lookup.price;
+    }
+
+    /**
+     * @dev We don't handle specifically the case where `ancillaryData` is not already readily translateable in utf8.
+     * For those cases, we assume that the client will be able to strip out the utf8-translateable part of the
+     * ancillary data that this contract stamps.
+     */
+    function _stampAncillaryData(bytes memory ancillaryData, address requester) internal pure returns (bytes memory) {
+        // Price requests that originate from this method, on Polygon, will ultimately be submitted to the DVM on
+        // Ethereum via the OracleRootTunnel. Therefore this contract should stamp its requester's address in the
+        // ancillary data so voters can conveniently track the requests path to the DVM.
+        return AncillaryData.appendAddressKey(ancillaryData, "childTunnelRequester", requester);
```

Comment:
Should we (in addition or as a replacement for the requester) stamp it with some sort of identifier for the chain (like the chain id if we have it)? The reason I mention this is that:
1. For a particular chain, it's not clear that a voter _needs_ to know which contract this request is coming from (this isn't provided on mainnet). 
2. A voter would likely be able to back this info out as long as they know which chain it's coming from. We should still provide it if it's needed for price resolution, but the requesting contract would probably provide it in that case?
3. Addresses aren't unique across chains, so this doesn't uniquely identify the requester. You can deploy the same contract at the same address on multiple chains (uniswap does this, for example).

Thoughts?

Reply:
added in [this commit](https://github.com/UMAprotocol/protocol/pull/3054/commits/bc713c39438ad40de375b3f1c2dab6d4f66da20f). As discussed offline, converting the uint `chainId` to utf8 bytes was not trivial FYI.